src/_pytest/python_api.py

```diff
@@ -527,8 +540,35 @@ def _is_numpy_array(obj):
 
 # builtin pytest.raises helper
 
+_E = TypeVar("_E", bound=BaseException)
+
+
+@overload
+def raises(
+    expected_exception: Union["Type[_E]", Tuple["Type[_E]", ...]],
```

Comment:
this annotation isn't quite correct, it should just be `Union["Type[BaseException]", Tuple["Type[BaseException]", ...]]`

right now it's only allowing a second tuple of the same type and so it is decaying to the base class:

```python
from typing import Type
from typing import TypeVar
from typing import Tuple
from typing import Union

E = TypeVar("E", bound=BaseException)

def raises(tp: Union[Type[E], Tuple[Type[E], ...]]) -> E : ...


tp1, tp2 = TypeError, AssertionError

reveal_type(raises(tp1))
reveal_type(raises((tp1, tp2)))
```

```console
$ mypy t.py
t.py:13: note: Revealed type is 'builtins.TypeError*'
t.py:14: note: Revealed type is 'builtins.Exception*'
```

Reply:
Ohhh right.

The optimal type for the exception given a tuple `(FileNotFoundError, InterruptedError)` is `Union[FileNotFoundError, InterruptedError]`, I'll try to see if I can make it work.

But if not, getting the base class (`IOError` in this example) is still better than `BaseException`, no?

And the most common case is the non-tuple case, which works out fine, I'd hate to lose that (inspecting the exception would be quite annoying with `BaseException`).

Reply:
In the absence of something like [TypeScript's mapped types](https://www.typescriptlang.org/docs/handbook/advanced-types.html#mapped-types), I found this hard to achieve.

I getting the base type for a tuple is not that bad, so I don't think this should block the PR. But if someone can improve it I'll be happy to include.